test/distributed/org/apache/cassandra/distributed/test/SSTableGenerationTest.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.distributed.test;
+
+import java.io.IOException;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import org.apache.cassandra.cql3.UntypedResultSet;
+import org.apache.cassandra.db.Directories;
+import org.apache.cassandra.db.Keyspace;
+import org.apache.cassandra.db.SystemKeyspace;
+import org.apache.cassandra.distributed.Cluster;
+import org.apache.cassandra.distributed.shared.ClusterUtils;
+import org.apache.cassandra.io.sstable.Descriptor;
+import org.apache.cassandra.io.sstable.SequenceBasedSSTableUniqueIdentifier;
+import org.apache.cassandra.io.sstable.UUIDBasedSSTableUniqueIdentifier;
+import org.apache.cassandra.metrics.RestorableMeter;
+import org.apache.cassandra.tools.SystemExitException;
+import org.apache.cassandra.utils.UUIDGen;
+import org.assertj.core.api.Assertions;
+import org.assertj.core.data.Offset;
+
+import static org.apache.cassandra.cql3.QueryProcessor.executeInternal;
+import static org.apache.cassandra.db.SystemKeyspace.LEGACY_SSTABLE_ACTIVITY;
+import static org.apache.cassandra.db.SystemKeyspace.SSTABLE_ACTIVITY_V2;
+import static org.apache.cassandra.distributed.shared.FutureUtils.waitOn;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class SSTableGenerationTest extends TestBaseImpl
```

Comment:
Maybe we should add similar JVM upgrade tests testing what happens when upgrading from older versions, wdyt?

Reply:
We can, though this test verifies what needs to be verified I think:
- `testRestartWithUUIDEnabled` starts a node with new ids feature disabled, creates some sstables and stops the node, those sstables are verified to conform the old naming format; then the node is restarted, thus this is very similar to upgrading the node where there are existing sstables with old numbering
- `testRestartWithUUIDDisabled` shows that we cannot disable the feature if we have existing sstables generated with UUID

I can add scenario where we just start a node with the feature disabled, create some sstables and restart the node (keep the feature disabled), but this scenario is well covered by many other tests

Reply:
What I would like to be sure is that we do not have some untested usecases that can lead to some issue in production. The scenarios that come to my mind are:
* streaming (zerocopy) between 2 nodes. One with the feature enabled the other one with the feature disabled
* compactions running with a mix of old names and new names (for the different compaction strategies)
* restore from a backup (old files only) with only new files or with a mix of new and old
* restore from a backup (old and new files)  with only new files or with a mix of new and old

Reply:
For the streaming scenario I would include a upgrade test with a pre-4.1 node and a 4.1 node with the feature enabled.

Reply:
tbh, I added simple streaming tests cases (streaming from a node with mixed sstables to a node with feature disabled and to a node with feature enabled) but adding more does not seem to be relevant - streaming has nothing to do with sstable file names - they are created locally by the stream receiver. Stream header does not transport any information about the identifier and on the receiver side, the sstable writer is always created with a new descriptor